crates/shared/src/request_sharing.rs

```diff
@@ -30,13 +36,32 @@ pub type BoxRequestSharing<Request, Response> =
 /// A boxed shared future.
 pub type BoxShared<T> = Shared<BoxFuture<'static, T>>;
 
-impl<Request, Fut: Future> RequestSharing<Request, Fut> {
+type Cache<Request, Response> = Arc<Mutex<HashMap<Request, WeakShared<Response>>>>;
+
+impl<Request: Send + 'static, Fut: Future + Send + 'static> RequestSharing<Request, Fut>
+where
+    Fut::Output: Send + Sync,
+{
     pub fn labelled(request_label: String) -> Self {
+        let cache: Cache<Request, Fut> = Default::default();
+        Self::spawn_gc(cache.clone());
         Self {
-            in_flight: Default::default(),
+            in_flight: cache,
             request_label,
         }
     }
+
+    fn collect_garbage(cache: &Cache<Request, Fut>) {
+        let mut cache = cache.lock().unwrap();
+        cache.retain(|_request, weak| weak.upgrade().is_some());
+    }
+
+    fn spawn_gc(cache: Cache<Request, Fut>) {
+        tokio::task::spawn(async move {
+            Self::collect_garbage(&cache);
+            tokio::time::sleep(Duration::from_millis(500)).await;
```

Comment:
Shouldn't this run in some form of loop? 

If we don't want to create separate task for this, we could probably also collect garbage during background updates in the recent block cache (but I guess other use cases would also like gc).

Reply:
> Shouldn't this run in some form of loop?

Absolutely, good catch!

> If we don't want to create separate task for this, we could probably also collect garbage during background updates in the recent block cache (but I guess other use cases would also like gc).

Yeah, other components that use `RequestSharing` definitely need a way to reclaim memory as well.